ami/models/hifigan_discriminators.py

```diff
@@ -0,0 +1,237 @@
+# Ref:
+# https://github.com/jik876/hifi-gan/blob/master/models.py
+
+import logging
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn.utils.parametrizations import spectral_norm, weight_norm
+
+
+def get_padding(kernel_size: int, dilation: int = 1) -> int:
+    return int((kernel_size * dilation - dilation) / 2)
+
+
+class DiscriminatorP(nn.Module):
+    """DiscriminatorP that constitutes MultiPeriodDiscriminator described
+    below.
+
+    Discriminate authenticity of the periodicity of the input audio.
+    """
+
+    def __init__(
+        self,
+        period: int,
+        in_channels: int = 1,
+        kernel_size: int = 5,
+        stride: int = 3,
+    ) -> None:
+        super().__init__()
+        self.period = period
+        self.convs = nn.ModuleList(
+            [
+                weight_norm(nn.Conv2d(in_channels, 32, (kernel_size, 1), (stride, 1), padding=(get_padding(5, 1), 0))),
+                weight_norm(nn.Conv2d(32, 128, (kernel_size, 1), (stride, 1), padding=(get_padding(5, 1), 0))),
+                weight_norm(nn.Conv2d(128, 512, (kernel_size, 1), (stride, 1), padding=(get_padding(5, 1), 0))),
+                weight_norm(nn.Conv2d(512, 1024, (kernel_size, 1), (stride, 1), padding=(get_padding(5, 1), 0))),
+                weight_norm(nn.Conv2d(1024, 1024, (kernel_size, 1), 1, padding=(2, 0))),
+            ]
+        )
```

Comment:
ここはConv1dで実装できる内容であること、またfor文を使って実装することが簡潔かつ健全です。
またマジックナンバーが多すぎます。ちゃんと変数にまとめ、説明的にしてください。
DRY (Don't Repeat Yourself)の原則に従い、繰り返し似た内容のコードを書くことは基本避けましょう。（正し、それによって大規模な修正が必要になる場合はこの限りではありません。将来的な修正項目にはなりますが）
```suggestion
        channels = [
            (in_channels, 32),
            (32, 128),
            (128, 512),
            (512, 1024),
            (1024, 1024)
        ]
        self.convs = nn.ModuleList()
        for i, (in_ch, out_ch) in enumerate(channels, start=1):
            padding = get_padding(kernel_size, 1)
            if i == len(channels):
                padding = 2
            self.convs.append(
                weight_norm(nn.Conv1d(in_ch, out_ch, kernel_size, stride, padding)
            )
```

Reply:
https://github.com/MLShukai/ami/pull/242/commits/461767c4bcbdea7e7ea028daa3aca932b5c213b9 で修正